doc/manual/modular-configurations.md

```diff
@@ -0,0 +1,429 @@
+composition: it pastes records together, the fields without definition in the
+result (the arguments) being the union of the fields without definition from
+each part.
+
+```nickel #parse
+{
+  all = network & services & volumes,
+  partial = network & services,
+}
+```
+
+## Toward modules
+
+Nickel is centered around records as the primary unit of configuration. Metadata
+can be attached to record fields, giving them more expressive power and the
+ability to better describe an _interface_ for a partial configuration. Combined
+with field metadata, you can use records - that is, partial configurations - as
+a reusable configuration module. For example:
+
+```nickel #parse
+{
+  inputs | not_exported = {
+      foo
+        | String
+        | doc "Doc of foo",
+      bar
+        | Number
+        | doc "Doc of bar",
+      unused
+        | Bool
+        | optional,
+    },
+  local | not_exported = {
+      computed = std.string.to_lower inputs.foo,
+    },
+
+  some_config_option = inputs.bar + 1,
+  other_option = std.string.join ["Hello", local.computed],
+  last_option = "values are %{local.computed} and %{std.to_string inputs.bar}",
+}
+```
+
+We used the `not_exported` metadata to signal that the `inputs` and `local`
+fields are high-level values that are used to generate the configuration, but
+they shouldn't be part of the final export. `inputs` is supposed to be set by
+consumers of this module. `local` contains intermediate values that
+are computed from the inputs and possibly reused many times in the outputs.
+
+Finally, the outputs lie at the root of the record. We could also have put them
+under an `output` field, requiring to extract the `output` field in the end
+(but, in return, we can then remove the `not_exported`).
+
+All of those values are readily and automatically overridable (using the `force`
+metadata for fields that already have a non-default value), and changes are
+propagated to reverse dependencies accordingly. The proposed layout is rather
+arbitrary: Nickel has no notion of inputs, local and outputs, but rather of
+fields with and without definition as well as priorities.
+
+The metadata associated to `inputs` define a clear typed module interface which
+can be leveraged by the Nickel tooling.
+
+Nickel might get a first-class notion of modules one day, if only to enforce a
+standard structure across projects. But, as far as functionality is concerned,
+the current merge system can already get you pretty far in an intuitive way.
+
+## Conclusion
+
+In this section, we explored the merge system of Nickel and how to leverage it
+to write natural configuration parts that are easy to query, to combine and to
+tweak. We supported our claim that partial configuration is a better suited
+model than functions (or other abstractions, for that matter) for configuration
+management, and showed that the combination of merging and field metadata is an
+expressive combo.
```

Comment:
I feel like we should remove this conclusion. It doesn't really fit the style of the other manual sections.

Reply:
Fair enough. I wondered about it too, because we have "take-away" subsections in some sections (typing I believe), but I agree it doesn't really fit.